File: Software/Python/line_follower/line_sensor.py
#!/usr/bin/env python
# Dexter Industries line sensor python library
#
# This library provides the basic functions to access the sensor data from the line sensor
#
# Have a question about this example?  Ask on the forums here:  http://www.dexterindustries.com/forum/?forum=grovepi
#
# Karan Nayan
# Initial Date: 13 Dec 2015
# Last Updated: 05 Apr 2017
# http://www.dexterindustries.com/
'''
## License
 Copyright (C) 2017  Dexter Industries

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/gpl-3.0.txt>.
'''
from __future__ import print_function
#from __future__ import division
from builtins import input
# the above lines are meant for Python3 compatibility.
# they force the use of Python3 functionality for print(),
# the integer division and input()
# mind your parentheses!

import time
import math
import RPi.GPIO as GPIO
import struct
import operator
import pickle
import numpy

from periphery import I2C, I2CError
# We are using [python-periphery] package because with the
# latest kernel versions (>v4.4.50-v7) and with the [python-smbus] package
# the line follower no longer works. The issue is with the line follower,
# which doesn't know how to process repeated starts from the master.
# With [python-periphery] package, it's possible to avoid issuing
# repeated start conditions and instead just use start/stop conditions.
# original report: https://github.com/raspberrypi/firmware/issues/828
# dev issue      : https://github.com/RobertLucian/GoPiGo/issues/8
# PR             : https://github.com/DexterInd/GoPiGo/pull/273


debug = 0

rev = GPIO.RPI_REVISION
if rev == 2 or rev == 3:
	bus_number = 1
else:
	bus_number = 0

#Add a multipler to each sensor
multp=[-10,-5,0,5,10]

#TRAIN for the first time
#reading when all sensors are at white
# white=[767,815,859,710,700]
white_line=[0]*5
#reading when all sensors are black
# black=[1012,1013,1015,1003,1004]
black_line=[0]*5
#difference between black and white
# range_col=list(map(operator.sub, black, white))
range_col=[0]*5

dir_path="/home/pi/Dexter/"
file_b=dir_path+'black_line.txt'
file_w=dir_path+'white_line.txt'
file_r=dir_path+'range_line.txt'
# Function declarations of the various functions used for encoding and sending
# data from RPi to Arduino

# buffer with raw data from the line follower sensor
sensor_buffer = [ [], [], [], [], [] ]
# keep a maximum of 20 readings for each IR sensor on the line follower
max_buffer_length = 20

# Function for removing outlier values
# For bigger std_factor_threshold, the filtering is less aggressive
# For smaller std_factor_threshold, the filtering is more aggressive
# std_factor_threshold must be bigger than 0
def statisticalNoiseReduction(values, std_factor_threshold = 2):
    if len(values) == 0:
    	return []

    mean = numpy.mean(values)
    standard_deviation = numpy.std(values)

    # just return if we only got constant values
    if standard_deviation == 0:
    	return values

    # remove outlier values which are less than the average but bigger than the calculated threshold
    filtered_values = [element for element in values if element > mean - std_factor_threshold * standard_deviation]
    # the same but in the opposite direction
    filtered_values = [element for element in filtered_values if element < mean + std_factor_threshold * standard_deviation]

    return filtered_values

# Function for reading line follower's values off of its IR sensor
def read_sensor():
    address = 0x06
    register = 0x01
    command = 0x03
    unused = 0x00

    try:
        i2c = I2C('/dev/i2c-' + str(bus_number))

        read_bytes = 10 * [0]
        msg1 = [ I2C.Message([register, command] + 3 * [unused]) ]
        msg2 = [ I2C.Message(read_bytes, read=True) ]
        # we meed to do 2 transfers so we can avoid using repeated starts
        # repeated starts don't go hand in hand with the line follower
        i2c.transfer(address, msg1)
        i2c.transfer(address, msg2)
        
    except I2CError as error:
        return 5 * [-1]

    # unpack bytes received and process them
    # bytes_list = struct.unpack('10B',read_results[0])
    output_values = []
    input_values = msg2[0].data

    for step in range(5):
        # calculate the 16-bit number we got
        sensor_buffer[step].append(input_values[2 * step] * 256 + input_values[2 * step + 1])

        # if there're too many elements in the list
        # then remove one
        if len(sensor_buffer[step]) > max_buffer_length:
            sensor_buffer[step].pop(0)

        # eliminate outlier values and select the most recent one
        filtered_value = statisticalNoiseReduction(sensor_buffer[step], 2)[-1]

        # append the value to the corresponding IR sensor
        output_values.append(filtered_value)

    return output_values

def get_sensorval():

	# updated to avoid an infinite loop
	attempt = 0
	while attempt < 5:
		val=read_sensor()
		# print (val)
		if val[0]!=-1:
			return val
		else:
			#Read once more to clear buffer and remove junk values
			val=read_sensor()
			attempt = attempt + 1

	return val


def set_black_line():
	global black_line, white_line, range_col
	for i in range(5):
		val=read_sensor()
	# print (val)
	if val[0]!=-1:
		black_line=val
	else:
		black_line=[-1]*5
	range_col=list(map(operator.sub, black_line, white_line))
	with open(file_b, 'wb') as f:
		pickle.dump(black_line, f)
	with open(file_r, 'wb') as f:
		pickle.dump(range_col, f)


def get_black_line():
	global black_line
	#load default values from files
	try:
		with open(file_b, 'rb') as f:
			black_line = pickle.load(f)
	except Exception as e:
		print ("Line Follower: failed getting black line values!")
		print (e)
		black_line=[0]*5
	return black_line


def set_white_line():
	global white_line,black_line,range_col
	for i in range(5):
		val=read_sensor()
	# print (val)
	if val[0]!=-1:
		white_line=val
	else:
		white_line=[-1]*5
	range_col=list(map(operator.sub, black_line, white_line))
	with open(file_w, 'wb') as f:
		pickle.dump(white_line, f)
	with open(file_r, 'wb') as f:
		pickle.dump(range_col, f)


def get_white_line():
	global white_line
	#load default values from files
	try:
		with open(file_w, 'rb') as f:
			white_line = pickle.load(f)
	except:
		white_line=[0]*5
	return white_line


def get_range():
	global range_col
	#load default values from files
	try:
		with open(file_r, 'rb') as f:
			range_col = pickle.load(f)
	except:
		range_col=[0]*5
	return range_col


def line_position():
	global black_line,white_line,range_col
	#load default values from files
	try:
		with open(file_b, 'rb') as f:
			black_line = pickle.load(f)
	except:
		black_line=[0]*5

	try:
		with open(file_w, 'rb') as f:
			white_line = pickle.load(f)
	except:
		white_line=[0]*5

	try:
		with open(file_r, 'rb') as f:
			range_col = pickle.load(f)
	except:
		range_col=[0]*5

	curr=get_sensorval()
	diff_val=list(map(operator.sub, curr, white_line))
	curr_pos=0
	percent_black_line=[0]*5
	for i in range(5):
		# casting to int when moving to Python 3
		percent_black_line[i]=(diff_val[i]*100/range_col[i])
		curr_pos+=percent_black_line[i]*multp[i]
	return curr_pos

File: Software/C/line_follower/line_sensor.c
// ####################################################################################
// Dexter Industries line sensor C library
//
//This library provides the basic functions to access the sensor data from the line sensor
//http://www.dexterindustries.com/GoPiGo/gopigo-line-follower-getting-started/
//
// History
// ------------------------------------------------
// Shoban Narayan
// Date               Comments
// 21 Sep 16          Initial Authoring

// ## License
//
// GoPiGo for the Raspberry Pi: an open source robotics platform for the Raspberry Pi.
// Copyright (C) 2017  Dexter Industries

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/gpl-3.0.txt>.

// ####################################################################################


#include "line_sensor.h"
 
int fd;
char *fileName = "/dev/i2c-1";
unsigned char w_buf[WRITE_BUF_SIZE],r_buf[READ_BUF_SIZE];
int read_val[5];
int i=0;
unsigned long reg_addr=0;

//Initialize
int init(void)
{
   
    if ((fd = open(fileName, O_RDWR)) < 0)
    {
        // Open port for reading and writing
        printf("Failed to open i2c port\n");
        return -1;
    }

    if (ioctl(fd, I2C_SLAVE, address) < 0)
    {
        // Set the port options and set the address of the device
        printf("Unable to get bus access to talk to slave\n");
        return -1;
    }
	
    return 1;
}

//Sleep in ms 
void sleep_ms(int t)
{
    usleep(t*1000);
}

//Write a register
int write_block(char cmd,char v1,char v2,char v3)
{
    w_buf[0]=1;
    w_buf[1]=cmd;
    w_buf[2]=v1;
    w_buf[3]=v2;
    w_buf[4]=v3;

    ssize_t ret = write(fd, w_buf, WRITE_BUF_SIZE);
    // sleep for 1 ms to prevent too fast writing
    sleep_ms(1);

    if (ret != WRITE_BUF_SIZE) {
        if (ret == -1) {
            printf("Error writing to GoPiGo (errno %i): %s\n", errno, strerror(errno));
        }
        else {
            printf("Error writing to GoPiGo\n");
        }
        return ret;
    }
    return 1;
}

//Read 1 byte of data
char read_byte(void)
{
    int reg_size=1;
    ssize_t ret = read(fd, r_buf, reg_size);

    if (ret != reg_size) {
        if (ret == -1) {
            printf("Unable to read from GoPiGo (errno %i): %s\n", errno, strerror(errno));
        }
        else {
            printf("Unable to read from GoPiGo\n");
        }
        return -1;
    }

    return r_buf[0];
}

// Read Line Sensor Values(each 10 Bit) to the buffer
int read_sensor(void)
{       int flag = 0;
	write_block(line_read_cmd,0,0,0); // This Write statement takes 0.895ms
	int reg_size=10;
	//Used to read the data from line sensor using i2c - read(int fildes, void *buf, size_t nbyte);
	ssize_t ret = read(fd, r_buf, reg_size);//This read statement takes 1.635ms to read from 5 IR sensors   
	if (ret != reg_size) {                   
        if (ret == -1) {
            printf("Unable to read from Line Sensor (errno %i): %s\n", errno, strerror(errno));
        }
        else {
            printf("Unable to read from Line_Sensor\n");
        }
        return -1;
    }
	for(i=0;i<10;i=i+2){                    // To convert the 10 bit analog reading of each sensor to decimal and store it in read_val[]
		read_val[i/2]=r_buf[i]*256+r_buf[i+1]; // Values less than 100 - White, Values greater than 800- Black
                if (read_val[i/2]> 65000)              // Checking for junk values in the input
			flag=1;
	}
        if (flag==1){
		for(i=0;i<5;i++)
			read_val[i]=-1;                    // Making junk input values to -1
        }        
	return 0;
}



